substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/DebugHelper.java

```diff
@@ -281,6 +295,493 @@ public static void printLocationInfo(@SuppressWarnings("unused") IsolateThread t
+            try {
+                return Word.objectToUntrackedPointer(Unsafe.getUnsafe().allocateInstance((Class)classObject));
+            } catch (InstantiationException e) {
+                return Word.objectToUntrackedPointer(null);
+            }
+        }
+
+        /**
+         * Pin object.
+         * @param thread
+         * @param ptrObj pointer to object to pin.
+         * @return "pin object" for given object, should be passed svm_dbg_unpin_object.
+         */
+        @CEntryPoint(name = "svm_dbg_pin_object", include = IncludeDebugHelperMethods.class, publishAs = Publish.SymbolOnly)
+        @CEntryPointOptions(prologue = SetThreadAndHeapBasePrologue.class)
+        public static Pointer pinObject(@SuppressWarnings("unused") IsolateThread thread, Pointer ptrObj) {
+            Object obj = ptrObj.toObject();
+            return Word.objectToUntrackedPointer(PinnedObject.create(obj));
+        }
+
+        /**
+         * Unpin object.
+         * @param thread
+         * @param ptrObj "pin object" @see svm_dbg_pin_object
+         */
+        @CEntryPoint(name = "svm_dbg_unpin_object", include = IncludeDebugHelperMethods.class, publishAs = Publish.SymbolOnly)
+        @CEntryPointOptions(prologue = SetThreadAndHeapBasePrologue.class)
+        public static void unPinObject(@SuppressWarnings("unused") IsolateThread thread, Pointer ptrObj) {
+            PinnedObject obj = (PinnedObject)ptrObj.toObject();
+            obj.close();
+        }
+
+        /**
+         * Returns modifier for given class.
+         * @param thread
+         * @param ptrClass pointer to class.
+         * @return integer representation for class's modifier.
+         */
+        @CEntryPoint(name = "svm_dbg_class_modifier", include = IncludeDebugHelperMethods.class, publishAs = Publish.SymbolOnly)
+        @CEntryPointOptions(prologue = SetThreadAndHeapBasePrologue.class)
+        public static int classGetModifier(@SuppressWarnings("unused") IsolateThread thread, Pointer ptrClass) {
+            Class<?> clazz = (Class<?>) ptrClass.toObject();
+            DynamicHub hub = DynamicHub.fromClass(clazz);
+            return hub.getModifiers();
+        }
+
+        /**
+         * Returns array of fields' descriptors.
+         * @param thread
+         * @param ptrClass pointer of class.
+         * @return array of fields descriptors in format name:modifier:type or name:modifier, depends on how much information
+         * Graal VM can provide.
+         */
+        @CEntryPoint(name = "svm_dbg_class_fields", include = IncludeDebugHelperMethods.class, publishAs = Publish.SymbolOnly)
+        @CEntryPointOptions(prologue = SetThreadAndHeapBasePrologue.class)
+        public static Pointer classGetFields(@SuppressWarnings("unused") IsolateThread thread, Pointer ptrClass) {
+            Class<?> clazz = (Class<?>)ptrClass.toObject();
+            DynamicHub hub = DynamicHub.fromClass(clazz);
+            Field[] fields = hub.getReachableFields();
```

Comment:
Most of the fields won't be available because this only includes the fields that are used via reflection. Same applies to other methods that access metadata.

Reply:
> Most of the fields won't be available because this only includes the fields that are used via reflection. Same applies to other methods that access metadata.

This is actually why we're have this method for fields and methods, to detect which fields and methods could be accessed from evaluator.

Reply:
If the heap dumping support is enabled, we have information about all the fields that are present in the Native Image (in a compact, encoded form but it should still be good enough for the debugging support).